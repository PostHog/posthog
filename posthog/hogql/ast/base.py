import re
from typing import Optional, Literal

from pydantic import BaseModel, Extra

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException

# NOTE: when you add new AST fields or nodes, add them to the Visitor classes in visitor.py as well!

camel_case_pattern = re.compile(r"(?<!^)(?=[A-Z])")


class AST(BaseModel):
    class Config:
        extra = Extra.forbid

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        from posthog.hogql.ast import UnknownType

        return UnknownType()


class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self


class Expr(AST):
    type: Optional[Type]


class Macro(Expr):
    name: str
    expr: Expr
    # Whether the macro is an inlined column "SELECT 1 AS a" or a subquery "SELECT a AS (SELECT 1)"
    macro_format: Literal["column", "subquery"]
