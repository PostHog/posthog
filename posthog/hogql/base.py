import re
from dataclasses import dataclass, field

from typing import Literal, Optional

from posthog.hogql.constants import ConstantDataType
from posthog.hogql.errors import NotImplementedException


# Given a string like "CorrectHorseBS", match the "H" and "B", so that we can convert this to "correct_horse_bs"
camel_case_pattern = re.compile(r"(?<!^)(?<![A-Z])(?=[A-Z])")


@dataclass(kw_only=True)
class AST:
    start: Optional[int] = field(default=None)
    end: Optional[int] = field(default=None)

    def accept(self, visitor):
        camel_case_name = camel_case_pattern.sub("_", self.__class__.__name__).lower()
        if "hog_qlx" in camel_case_name:
            camel_case_name = camel_case_name.replace("hog_qlx", "hogqlx_")
        method_name = f"visit_{camel_case_name}"
        if hasattr(visitor, method_name):
            visit = getattr(visitor, method_name)
            return visit(self)
        if hasattr(visitor, "visit_unknown"):
            return visitor.visit_unknown(self)
        raise NotImplementedException(f"Visitor has no method {method_name}")


@dataclass(kw_only=True)
class Type(AST):
    def get_child(self, name: str) -> "Type":
        raise NotImplementedException("Type.get_child not overridden")

    def has_child(self, name: str) -> bool:
        return self.get_child(name) is not None

    def resolve_constant_type(self) -> Optional["ConstantType"]:
        return UnknownType()


@dataclass(kw_only=True)
class Expr(AST):
    type: Optional[Type] = field(default=None)


@dataclass(kw_only=True)
class CTE(Expr):
    """A common table expression."""

    name: str
    expr: Expr
    # Whether the CTE is an inlined column "WITH 1 AS a" or a subquery "WITH a AS (SELECT 1)"
    cte_type: Literal["column", "subquery"]


@dataclass(kw_only=True)
class ConstantType(Type):
    data_type: ConstantDataType

    def resolve_constant_type(self) -> "ConstantType":
        return self

    def print_type(self) -> str:
        raise NotImplementedException("ConstantType.print_type not implemented")


@dataclass(kw_only=True)
class UnknownType(ConstantType):
    data_type: ConstantDataType = field(default="unknown", init=False)

    def print_type(self) -> str:
        return "Unknown"
