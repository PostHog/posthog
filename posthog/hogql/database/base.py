from typing import Any, Callable, Dict, List

from pydantic import BaseModel, Extra

from posthog.hogql.errors import HogQLException, NotImplementedException
from posthog.hogql.database.fields import DatabaseField


class Table(BaseModel):
    class Config:
        extra = Extra.forbid

    def has_field(self, name: str) -> bool:
        return hasattr(self, name)

    def get_field(self, name: str) -> DatabaseField:
        if self.has_field(name):
            return getattr(self, name)
        raise HogQLException(f'Field "{name}" not found on table {self.__class__.__name__}')

    def clickhouse_table(self):
        raise NotImplementedException("Table.clickhouse_table not overridden")

    def hogql_table(self):
        raise NotImplementedException("Table.hogql_table not overridden")

    def avoid_asterisk_fields(self) -> List[str]:
        return []

    def get_asterisk(self) -> Dict[str, DatabaseField]:
        asterisk: Dict[str, DatabaseField] = {}
        fields_to_avoid = self.avoid_asterisk_fields() + ["team_id"]
        for key, field in self.__fields__.items():
            if key in fields_to_avoid:
                continue
            database_field = field.default
            if isinstance(database_field, DatabaseField):
                asterisk[key] = database_field
            elif (
                isinstance(database_field, Table)
                or isinstance(database_field, LazyJoin)
                or isinstance(database_field, FieldTraverser)
            ):
                pass  # ignore virtual tables for now
            else:
                raise HogQLException(f"Unknown field type {type(database_field).__name__} for asterisk")
        return asterisk


class LazyJoin(BaseModel):
    class Config:
        extra = Extra.forbid

    join_function: Callable[[str, str, Dict[str, Any]], Any]
    join_table: Table
    from_field: str


class LazyTable(Table):
    class Config:
        extra = Extra.forbid

    def lazy_select(self, requested_fields: Dict[str, Any]) -> Any:
        raise NotImplementedException("LazyTable.lazy_select not overridden")


class VirtualTable(Table):
    class Config:
        extra = Extra.forbid


class FieldTraverser(BaseModel):
    class Config:
        extra = Extra.forbid

    chain: List[str]
