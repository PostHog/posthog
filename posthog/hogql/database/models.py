from typing import Any, Callable, Dict, List, Optional, TYPE_CHECKING
from pydantic import ConfigDict, BaseModel

from posthog.hogql.errors import HogQLException, NotImplementedException
from posthog.schema import HogQLQueryModifiers

if TYPE_CHECKING:
    from posthog.hogql.context import HogQLContext
    from posthog.hogql.ast import SelectQuery


class FieldOrTable(BaseModel):
    pass


class DatabaseField(FieldOrTable):
    """
    Base class for a field in a database table.
    """

    model_config = ConfigDict(extra="forbid")

    name: str
    array: Optional[bool] = None
    nullable: Optional[bool] = None


class IntegerDatabaseField(DatabaseField):
    pass


class FloatDatabaseField(DatabaseField):
    pass


class StringDatabaseField(DatabaseField):
    pass


class StringJSONDatabaseField(DatabaseField):
    pass


class StringArrayDatabaseField(DatabaseField):
    pass


class DateDatabaseField(DatabaseField):
    pass


class DateTimeDatabaseField(DatabaseField):
    pass


class BooleanDatabaseField(DatabaseField):
    pass


class FieldTraverser(FieldOrTable):
    model_config = ConfigDict(extra="forbid")

    chain: List[str]


class Table(FieldOrTable):
    fields: Dict[str, FieldOrTable]
    model_config = ConfigDict(extra="forbid")

    def has_field(self, name: str) -> bool:
        return name in self.fields

    def get_field(self, name: str) -> FieldOrTable:
        if self.has_field(name):
            return self.fields[name]
        raise Exception(f'Field "{name}" not found on table {self.__class__.__name__}')

    def to_printed_clickhouse(self, context: "HogQLContext") -> str:
        raise NotImplementedException("Table.to_printed_clickhouse not overridden")

    def to_printed_hogql(self) -> str:
        raise NotImplementedException("Table.to_printed_hogql not overridden")

    def avoid_asterisk_fields(self) -> List[str]:
        return []

    def get_asterisk(self):
        fields_to_avoid = self.avoid_asterisk_fields() + ["team_id"]
        asterisk: Dict[str, FieldOrTable] = {}
        for key, field in self.fields.items():
            if key in fields_to_avoid:
                continue
            if isinstance(field, DatabaseField):
                asterisk[key] = field
            elif isinstance(field, Table) or isinstance(field, LazyJoin) or isinstance(field, FieldTraverser):
                pass  # ignore virtual tables for now
            else:
                raise HogQLException(f"Unknown field type {type(field).__name__} for asterisk")
        return asterisk


class LazyJoin(FieldOrTable):
    model_config = ConfigDict(extra="forbid")

    join_function: Callable[[str, str, Dict[str, Any], "HogQLContext", "SelectQuery"], Any]
    join_table: Table
    from_field: str


class LazyTable(Table):
    """
    A table that is replaced with a subquery returned from `lazy_select(requested_fields: Dict[name, chain], modifiers: HogQLQueryModifiers)`
    """

    model_config = ConfigDict(extra="forbid")

    def lazy_select(self, requested_fields: Dict[str, List[str]], modifiers: HogQLQueryModifiers) -> Any:
        raise NotImplementedException("LazyTable.lazy_select not overridden")


class VirtualTable(Table):
    """
    A nested table that reuses the parent for storage. E.g. events.person.* fields with PoE enabled.
    """

    model_config = ConfigDict(extra="forbid")


class FunctionCallTable(Table):
    """
    A table that returns a function call, e.g. numbers(...) or s3(...). The team_id guard is NOT added for these.
    """

    name: str
    min_args: Optional[int] = None
    max_args: Optional[int] = None


class SavedQuery(Table):
    """
    A table that returns a subquery, e.g. my_saved_query -> (SELECT * FROM some_saved_table). The team_id guard is NOT added for the overall subquery
    """

    query: str
    name: str

    # Note: redundancy for safety. This validation is used in the data model already
    def to_printed_clickhouse(self, context):
        from posthog.warehouse.models import validate_saved_query_name

        validate_saved_query_name(self.name)
        return self.name

    def to_printed_hogql(self):
        from posthog.warehouse.models import validate_saved_query_name

        validate_saved_query_name(self.name)
        return self.name
