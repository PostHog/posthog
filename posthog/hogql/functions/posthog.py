from posthog.hogql.ast import ArrayType, BooleanType, DateTimeType, DateType, DecimalType, IntegerType, StringType

from .core import HogQLFunctionMeta

HOGQL_POSTHOG_FUNCTIONS: dict[str, HogQLFunctionMeta] = {
    "matchesAction": HogQLFunctionMeta("matchesAction", 1, 1),
    "sparkline": HogQLFunctionMeta("sparkline", 1, 1),
    "recordingButton": HogQLFunctionMeta("recordingButton", 1, 2),
    "explainCSPReport": HogQLFunctionMeta("explainCSPReport", 1, 1),
    # Allow case-insensitive matching since people might not know "SemVer" is the right capitalization
    "sortablesemver": HogQLFunctionMeta(
        "arrayMap(x -> toInt64OrZero(x),  splitByChar('.', extract(assumeNotNull({}), '(\\d+(\\.\\d+)+)')))",
        1,
        1,
        case_sensitive=False,
        signatures=[((StringType(),), ArrayType(item_type=IntegerType()))],
    ),
    "embedText": HogQLFunctionMeta("embedText", 1, 2),
    # posthog/models/channel_type/sql.py and posthog/hogql/database/schema/channel_type.py
    "lookupDomainType": HogQLFunctionMeta("lookupDomainType", 1, 1),
    "lookupPaidSourceType": HogQLFunctionMeta("lookupPaidSourceType", 1, 1),
    "lookupPaidMediumType": HogQLFunctionMeta("lookupPaidMediumType", 1, 1),
    "lookupOrganicSourceType": HogQLFunctionMeta("lookupOrganicSourceType", 1, 1),
    "lookupOrganicMediumType": HogQLFunctionMeta("lookupOrganicMediumType", 1, 1),
    # posthog/models/exchange_rate/sql.py
    # convertCurrency(from_currency, to_currency, amount, timestamp?)
    "convertCurrency": HogQLFunctionMeta(
        "convertCurrency",
        3,
        4,
        signatures=[
            (
                (
                    StringType(),
                    StringType(),
                    DecimalType(),
                ),
                DecimalType(),
            ),
            (
                (
                    StringType(),
                    StringType(),
                    DecimalType(),
                    DateType(),
                ),
                DecimalType(),
            ),
            (
                (
                    StringType(),
                    StringType(),
                    DecimalType(),
                    DateTimeType(),
                ),
                DecimalType(),
            ),
        ],
    ),
    # survey functions
    "getSurveyResponse": HogQLFunctionMeta(
        "getSurveyResponse", 1, 3, signatures=[((IntegerType(), StringType(), BooleanType()), StringType())]
    ),
    "uniqueSurveySubmissionsFilter": HogQLFunctionMeta(
        "uniqueSurveySubmissionsFilter", 1, 1, signatures=[((StringType(),), StringType())]
    ),
}
