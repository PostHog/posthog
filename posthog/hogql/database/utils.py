from typing import Optional

from posthog.hogql import ast
from posthog.hogql.parser import parse_expr

from posthog.exceptions_capture import capture_exception


def get_join_field_chain(key: str) -> Optional[list[str | int]]:
    field = parse_expr(key)
    if isinstance(field, ast.Field):
        return field.chain

    if isinstance(field, ast.Alias) and isinstance(field.expr, ast.Call) and isinstance(field.expr.args[0], ast.Field):
        return field.expr.args[0].chain

    if isinstance(field, ast.Call) and isinstance(field.args[0], ast.Field):
        return field.args[0].chain

    capture_exception(Exception(f"Data Warehouse Join HogQL expression should be a Field or Call node: {key}"))
    return None
