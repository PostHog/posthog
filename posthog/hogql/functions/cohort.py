from typing import List

from posthog.hogql import ast
from posthog.hogql.context import HogQLContext
from posthog.hogql.errors import HogQLException
from posthog.hogql.escape_sql import escape_clickhouse_string
from posthog.hogql.parser import parse_expr


def cohort_subquery(cohort_id, is_static) -> ast.Expr:
    if is_static:
        sql = "(SELECT person_id FROM static_cohort_people WHERE cohort_id = {cohort_id})"
    else:
        sql = "(SELECT person_id FROM raw_cohort_people WHERE cohort_id = {cohort_id} GROUP BY person_id, cohort_id, version HAVING sum(sign) > 0)"
    return parse_expr(sql, {"cohort_id": ast.Constant(value=cohort_id)}, start=None)  # clear the source start position


def cohort_query_node(node: ast.Expr, context: HogQLContext) -> ast.Expr:
    return cohort(node, [node], context)


def cohort(node: ast.Expr, args: List[ast.Expr], context: HogQLContext) -> ast.Expr:
    arg = args[0]
    if not isinstance(arg, ast.Constant):
        raise HogQLException("cohort() takes only constant arguments", node=arg)

    from posthog.models import Cohort

    if isinstance(arg.value, int) and not isinstance(arg.value, bool):
        cohorts = Cohort.objects.filter(id=arg.value, team_id=context.team_id).values_list("id", "is_static", "name")
        if len(cohorts) == 1:
            context.add_notice(
                start=arg.start,
                end=arg.end,
                message=f"Cohort #{cohorts[0][0]} can also be specified as {escape_clickhouse_string(cohorts[0][2])}",
                fix=escape_clickhouse_string(cohorts[0][2]),
            )
            return cohort_subquery(cohorts[0][0], cohorts[0][1])
        raise HogQLException(f"Could not find cohort with id {arg.value}", node=arg)

    if isinstance(arg.value, str):
        cohorts = Cohort.objects.filter(name=arg.value, team_id=context.team_id).values_list("id", "is_static")
        if len(cohorts) == 1:
            context.add_notice(
                start=arg.start,
                end=arg.end,
                message=f"Searching for cohort by name. Replace with numeric ID {cohorts[0][0]} to protect against renaming.",
                fix=str(cohorts[0][0]),
            )
            return cohort_subquery(cohorts[0][0], cohorts[0][1])
        elif len(cohorts) > 1:
            raise HogQLException(f"Found multiple cohorts with name '{arg.value}'", node=arg)
        raise HogQLException(f"Could not find a cohort with the name '{arg.value}'", node=arg)

    raise HogQLException("cohort() takes exactly one string or integer argument", node=arg)
