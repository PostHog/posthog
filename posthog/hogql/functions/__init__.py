from .mapping import (
    validate_function_args,
    HogQLFunctionMeta,
    HOGQL_CLICKHOUSE_FUNCTIONS,
    HOGQL_AGGREGATIONS,
    HOGQL_POSTHOG_FUNCTIONS,
    ADD_OR_NULL_DATETIME_FUNCTIONS,
    FIRST_ARG_DATETIME_FUNCTIONS,
)
from .cohort import cohort
from .sparkline import sparkline

__all__ = [
    "validate_function_args",
    "HogQLFunctionMeta",
    "HOGQL_CLICKHOUSE_FUNCTIONS",
    "HOGQL_AGGREGATIONS",
    "HOGQL_POSTHOG_FUNCTIONS",
    "ADD_OR_NULL_DATETIME_FUNCTIONS",
    "FIRST_ARG_DATETIME_FUNCTIONS",
    "cohort",
    "sparkline",
]
