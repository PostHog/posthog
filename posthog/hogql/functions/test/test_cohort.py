from django.test import override_settings

from posthog.hogql import ast
from posthog.hogql.errors import HogQLException
from posthog.hogql.parser import parse_expr
from posthog.hogql.query import execute_hogql_query
from posthog.models import Cohort
from posthog.models.cohort.util import recalculate_cohortpeople
from posthog.models.utils import UUIDT
from posthog.schema import HogQLQueryModifiers
from posthog.test.base import (
    BaseTest,
    _create_person,
    _create_event,
    flush_persons_and_events,
)

elements_chain_match = lambda x: parse_expr("match(elements_chain, {regex})", {"regex": ast.Constant(value=str(x))})
not_call = lambda x: ast.Call(name="not", args=[x])


class TestCohort(BaseTest):
    maxDiff = None

    def _create_random_events(self) -> str:
        random_uuid = str(UUIDT())
        _create_person(
            properties={"$os": "Chrome", "random_uuid": random_uuid},
            team=self.team,
            distinct_ids=["bla"],
            is_identified=True,
        )
        _create_event(distinct_id="bla", event=random_uuid, team=self.team)
        flush_persons_and_events()
        return random_uuid

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=True, PERSON_ON_EVENTS_V2_OVERRIDE=True)
    def test_in_cohort_dynamic(self):
        random_uuid = self._create_random_events()
        cohort = Cohort.objects.create(
            team=self.team,
            groups=[{"properties": [{"key": "$os", "value": "Chrome", "type": "person"}]}],
        )
        recalculate_cohortpeople(cohort, pending_version=0)
        response = execute_hogql_query(
            f"SELECT event FROM events WHERE person_id IN COHORT {cohort.pk} AND event='{random_uuid}'",
            self.team,
            modifiers=HogQLQueryModifiers(inCohortVia="subquery"),
        )
        self.assertEqual(
            response.clickhouse,
            f"SELECT events.event FROM events WHERE and(equals(events.team_id, {self.team.pk}), in(events.person_id, (SELECT cohortpeople.person_id FROM cohortpeople WHERE and(equals(cohortpeople.team_id, {self.team.pk}), equals(cohortpeople.cohort_id, {cohort.pk})) GROUP BY cohortpeople.person_id, cohortpeople.cohort_id, cohortpeople.version HAVING ifNull(greater(sum(cohortpeople.sign), 0), 0))), equals(events.event, %(hogql_val_0)s)) LIMIT 100 SETTINGS readonly=2, max_execution_time=60, allow_experimental_object_type=1",
        )
        self.assertEqual(
            response.hogql,
            f"SELECT event FROM events WHERE and(in(person_id, (SELECT person_id FROM raw_cohort_people WHERE equals(cohort_id, {cohort.pk}) GROUP BY person_id, cohort_id, version HAVING greater(sum(sign), 0))), equals(event, '{random_uuid}')) LIMIT 100",
        )
        self.assertEqual(len(response.results), 1)
        self.assertEqual(response.results[0][0], random_uuid)

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=True, PERSON_ON_EVENTS_V2_OVERRIDE=True)
    def test_in_cohort_static(self):
        cohort = Cohort.objects.create(
            team=self.team,
            is_static=True,
        )
        response = execute_hogql_query(
            f"SELECT event FROM events WHERE person_id IN COHORT {cohort.pk}",
            self.team,
            modifiers=HogQLQueryModifiers(inCohortVia="subquery"),
        )
        self.assertEqual(
            response.clickhouse,
            f"SELECT events.event FROM events WHERE and(equals(events.team_id, {self.team.pk}), in(events.person_id, (SELECT person_static_cohort.person_id FROM person_static_cohort WHERE and(equals(person_static_cohort.team_id, {self.team.pk}), equals(person_static_cohort.cohort_id, {cohort.pk}))))) LIMIT 100 SETTINGS readonly=2, max_execution_time=60, allow_experimental_object_type=1",
        )
        self.assertEqual(
            response.hogql,
            f"SELECT event FROM events WHERE in(person_id, (SELECT person_id FROM static_cohort_people WHERE equals(cohort_id, {cohort.pk}))) LIMIT 100",
        )

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=True, PERSON_ON_EVENTS_V2_OVERRIDE=True)
    def test_in_cohort_strings(self):
        cohort = Cohort.objects.create(
            team=self.team,
            name="my cohort",
            is_static=True,
        )
        response = execute_hogql_query(
            f"SELECT event FROM events WHERE person_id IN COHORT 'my cohort'",
            self.team,
            modifiers=HogQLQueryModifiers(inCohortVia="subquery"),
        )
        self.assertEqual(
            response.clickhouse,
            f"SELECT events.event FROM events WHERE and(equals(events.team_id, {self.team.pk}), in(events.person_id, (SELECT person_static_cohort.person_id FROM person_static_cohort WHERE and(equals(person_static_cohort.team_id, {self.team.pk}), equals(person_static_cohort.cohort_id, {cohort.pk}))))) LIMIT 100 SETTINGS readonly=2, max_execution_time=60, allow_experimental_object_type=1",
        )
        self.assertEqual(
            response.hogql,
            f"SELECT event FROM events WHERE in(person_id, (SELECT person_id FROM static_cohort_people WHERE equals(cohort_id, {cohort.pk}))) LIMIT 100",
        )

    @override_settings(PERSON_ON_EVENTS_OVERRIDE=True, PERSON_ON_EVENTS_V2_OVERRIDE=True)
    def test_in_cohort_error(self):
        with self.assertRaises(HogQLException) as e:
            execute_hogql_query(f"SELECT event FROM events WHERE person_id IN COHORT true", self.team)
        self.assertEqual(str(e.exception), "cohort() takes exactly one string or integer argument")

        with self.assertRaises(HogQLException) as e:
            execute_hogql_query(
                f"SELECT event FROM events WHERE person_id IN COHORT 'blabla'",
                self.team,
            )
        self.assertEqual(str(e.exception), "Could not find a cohort with the name 'blabla'")
