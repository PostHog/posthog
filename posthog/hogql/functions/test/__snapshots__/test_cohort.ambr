# name: TestCohort.test_in_cohort_dynamic
  '
  -- ClickHouse
  
  SELECT events.event AS event 
  FROM events 
  WHERE and(equals(events.team_id, 420), in(events.person_id, (
  SELECT cohortpeople.person_id AS person_id 
  FROM cohortpeople 
  WHERE and(equals(cohortpeople.team_id, 420), equals(cohortpeople.cohort_id, 6)) 
  GROUP BY cohortpeople.person_id, cohortpeople.cohort_id, cohortpeople.version 
  HAVING ifNull(greater(sum(cohortpeople.sign), 0), 0))), equals(events.event, %(hogql_val_0)s)) 
  LIMIT 100 
  SETTINGS readonly=2, max_execution_time=60, allow_experimental_object_type=1
  
  -- HogQL
  
  SELECT event 
  FROM events 
  WHERE and(in(person_id, (
  SELECT person_id 
  FROM raw_cohort_people 
  WHERE equals(cohort_id, 6) 
  GROUP BY person_id, cohort_id, version 
  HAVING greater(sum(sign), 0))), equals(event, 'RANDOM_TEST_ID::UUID')) 
  LIMIT 100
  '
---
# name: TestCohort.test_in_cohort_static
  '
  -- ClickHouse
  
  SELECT events.event AS event 
  FROM events 
  WHERE and(equals(events.team_id, 420), in(events.person_id, (
  SELECT person_static_cohort.person_id AS person_id 
  FROM person_static_cohort 
  WHERE and(equals(person_static_cohort.team_id, 420), equals(person_static_cohort.cohort_id, 7))))) 
  LIMIT 100 
  SETTINGS readonly=2, max_execution_time=60, allow_experimental_object_type=1
  
  -- HogQL
  
  SELECT event 
  FROM events 
  WHERE in(person_id, (
  SELECT person_id 
  FROM static_cohort_people 
  WHERE equals(cohort_id, 7))) 
  LIMIT 100
  '
---
# name: TestCohort.test_in_cohort_strings
  '
  -- ClickHouse
  
  SELECT events.event AS event 
  FROM events 
  WHERE and(equals(events.team_id, 420), in(events.person_id, (
  SELECT person_static_cohort.person_id AS person_id 
  FROM person_static_cohort 
  WHERE and(equals(person_static_cohort.team_id, 420), equals(person_static_cohort.cohort_id, 8))))) 
  LIMIT 100 
  SETTINGS readonly=2, max_execution_time=60, allow_experimental_object_type=1
  
  -- HogQL
  
  SELECT event 
  FROM events 
  WHERE in(person_id, (
  SELECT person_id 
  FROM static_cohort_people 
  WHERE equals(cohort_id, 8))) 
  LIMIT 100
  '
---
